### ISSUE TITLE
cue/load: add a language version field to cue.mod

### ISSUE THREAD

**Issue by myitcv** [cue-lang/cue#2004]
A `go.mod` file can include a [go directive](https://go.dev/ref/mod#go-mod-file-go). Quoting the modules reference:

> The go directive was originally intended to support backward incompatible changes to the Go language (see [Go 2 transition](https://go.googlesource.com/proposal/+/master/design/28221-go2-transitions.md)). There have been no incompatible language changes since modules were introduced, but the go directive still affects use of new language features:
> 
> * For packages within the module, the compiler rejects use of language features introduced after the version specified by the go directive. For example, if a module has the directive go 1.12, its packages may not use numeric literals like 1_000_000, which were introduced in Go 1.13.
> * If an older Go version builds one of the module’s packages and encounters a compile error, the error notes that the module was written for a newer Go version. For example, suppose a module has go 1.13 and a package uses the numeric literal 1_000_000. If that package is built with Go 1.12, the compiler notes that the code is written for Go 1.13.

https://github.com/golang/go/discussions/55092 and https://github.com/golang/go/discussions/55090 discuss ways in which (language) changes can be made so as to retain backwards and forwards compatibility. 

With a [number of pre-v1.0 changes](https://github.com/cue-lang/cue/labels/pre%20v1.0) introducing breaking changes of one sort or another, having some means to control the version (semantics) of CUE that is used (from either a language or tooling perspective) in a `cue.mod` file will help with transitions. 

Creating this issue as a placeholder for that implementation.

Although this is not exclusively related to language changes, marking as such for visibility.

**Comment by mvdan** [comment]
https://github.com/golang/go/discussions/55092 proposes a `toolchain` line, then a `go.mod` would end up like:

```
go 1.91
toolchain go1.92.4
```

It's entirely possible that in the future we'll want to record a "toolchain" or "target" version to build with, akin to what Go did. Moreover, even today the `go` directive is confusing to many. It's only meant to track the _language_ version, but some also think it enforces other rules, like the minimum version required.

For both of those reasons, I suggest that we use `language` as the directive rather than `cue`, for example `language 0.6`. Then it's abundantly clear that it primarily affects what version of the language spec it's written for. I also don't think we need to explicitly use `cue` in the name, given that we're inside `cue.mod` already.

**Comment by mvdan** [comment]
developing this a bit further, how about:

```
version: language: "0.5"
```

Then, if we wanted to add a toolchain line in the future, it would fit just fine:

```
version: {
  language: "0.5"
  toolchain: "0.5.3"
}
```

**Comment by mvdan** [comment]
Spoke about this with @mpvl and @rogpeppe, and we think we can delay adding this field to the schema and automatically filling it in until v0.7. v0.6 doesn't strictly need the field for any logic around its language features or changes; it is adding required fields, but we're doing so unconditionally.

For the change in v0.7, we've agreed to reverse the fields, like `language: version: "v0.7"`. This makes the words make more intuitive sense - "language version" rather than "version language" - and it also feels more appropriate, as the version is a property of the language, and not the other way around.

We'll also automatically insert the version as part of `cue fix`. For example, if I upgrade from v0.6 to v0.7 and run `cue fix`, after it converts my CUE code to the latter version, it would insert or replace my `language.version` field to be `"v0.7"`. In the absence of any existing `language.version` field, `cue fix` could also infer the current version via heuristics based on the syntax and language features currently being used. `cue fix` already has to do those sort of heuristics to "fix" those old bits of code, so it doesn't feel like a lot of added work.

And, as a general fallback, a missing `language.version` field would mean "CUE language v0.6", since as of v0.6 we haven't really removed or broken any language features for quite a while. Akin to how Go treated a missing `go` line in `go.mod` to mean `go 1.11`.

**Comment by mvdan** [comment]
As of October 2023, CUE master already has an optional field `language.version`:

https://github.com/cue-lang/cue/blob/eddccfc72316a648971d9684ac5dcd003ee26385/internal/mod/modfile/schema.cue#L29-L35

One missing piece now would be for `cue fix ./...` to insert the current version if there's no such version - otherwise I suspect almost noone would be inserting these versions by themselves.

We should also start gating new language features or changes based on this version. For example, if v0.8 were to add a new builtin, it would only be available if `semver.Compare(language.version, "v0.8") >= 0`.

@rogpeppe looking at your strict module schema, I see that `language.version` is a `#Semver`. This means we would accept e.g. `v0.8.0` but presumably reject `v0.8`. Personally I wonder if the language version should just be `MAJOR.MINOR` (per the semver spec) as opposed to `MAJOR.MINOR.PATCH`, since we should never be adding features or making major changes to the language spec in patch releases. This would also match Go, where most people write e.g. `go 1.21` and not `go 1.21.0` - they only use the patch version when specifying a particular toolchain, where the entire semver does matter.

**Comment by mvdan** [comment]
This was implemented for both `cue mod tidy` and `cue mod init` by February 2024, and released as part of v0.8.0 in March - see https://github.com/cue-lang/cue/releases/tag/v0.8.0:

> [CL 1176194](https://cuelang.org/cl/1176194) adds a `language.version` field to `cue.mod/module.cue` to track what CUE language version a module was written for. This will become necessary to make future language changes as smooth as possible for CUE users. For example, running `cue mod init` or `cue mod tidy` with the v0.8.0 release should add `language: version: "v0.8.0"` when the field isn't present.
